bundles/org.openhab.binding.pentair/src/main/java/org/openhab/binding/pentair/internal/handler/PentairControllerHandler.java

```diff
@@ -0,0 +1,1134 @@
+
+    public PentairControllerHandler(Thing thing) {
+        super(thing);
+
+        for (int i = 0; i < NUM_SCHEDULES; i++) {
+            schedulesCache[i] = new ExpiringCache<>(CACHE_EXPIRY_LONG);
+        }
+
+        for (int i = 0; i < NUM_CIRCUITS; i++) {
+            circuitsCache[i] = new ExpiringCache<>(CACHE_EXPIRY_LONG);
+        }
+    }
+
+    @Override
+    public void readConfiguration() {
+        PentairControllerHandlerConfig config = getConfigAs(PentairControllerHandlerConfig.class);
+
+        this.id = config.id;
+        this.syncTime = config.synctime;
+    }
+
+    @Override
+    public void goOnline() {
+        // Only a single controller is supported on the Pentair bus so prevent multiple controller
+        // things being created.
+        if (onlineController != null) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "Another controller is already configured.");
+        } else {
+            super.goOnline();
+        }
+    }
+
+    @Override
+    public void finishOnline() {
+        super.finishOnline();
+
+        // setup syncTimeJob to run once a day, initial time to sync is 3 minutes after controller goes online. This is
+        // to prevent collision with main thread queries on initial startup
+        syncTimeJob = scheduler.scheduleWithFixedDelay(() -> syncTime(), 3, 24 * 60 * 60, TimeUnit.MINUTES);
+
+        scheduler.execute(() -> readControllerSettings());
+    }
+
+    public void syncTime() {
+        boolean synctime = ((boolean) getConfig().get("synctime"));
+        if (synctime) {
+            logger.debug("Synchronizing System Time with Pentair controller");
+            Calendar now = Calendar.getInstance();
+            sendClockSettings(now.get(Calendar.HOUR_OF_DAY), now.get(Calendar.MINUTE), now.get(Calendar.DAY_OF_WEEK),
+                    now.get(Calendar.DAY_OF_MONTH), now.get(Calendar.MONTH) + 1, now.get(Calendar.YEAR) - 2000);
+        }
+    }
+
+    public void readControllerSettings() {
+        int i;
+
+        requestSWVersion();
+        requestHeatStatus();
+        requestClockSettings();
+
+        for (i = 1; i <= NUM_CIRCUITS; i++) {
+            requestCircuitNameFunction(i);
+        }
+
+        for (i = 1; i <= NUM_SCHEDULES; i++) {
+            requestSchedule(i);
+        }
+
+        requestLightGroups();
+        requestValves();
+    }
+
+    @Override
+    public void goOffline(ThingStatusDetail detail) {
+        super.goOffline(detail);
+
+        if (syncTimeJob != null) {
+            syncTimeJob.cancel(true);
+        }
+
+        onlineController = null;
+    }
+
+    public @Nullable PentairControllerCircuit getCircuitByGroupID(String groupID) {
+        for (ExpiringCache<PentairControllerCircuit> circuitItem : circuitsCache) {
+            PentairControllerCircuit circuit = circuitItem.getLastKnownValue();
+            if (circuit == null) {
+                break;
+            }
```

Comment:
Why should this break? Should it continue instead?
```suggestion
            if (circuit == null) {
                continue;
            }
```

Reply:
yes - good catch.